Activator

ryssbowh opened the pull request:
Hi,

as discussed here : https://github.com/nWidart/laravel-modules/issues/789

mikemand commented on config/config.php:
```diff
@@ -166,7 +168,7 @@
         'enabled' => false,
         'key' => 'laravel-modules',
-        'lifetime' => 60,
+        'lifetime' => 60
```
Why? Trailing commas is pretty much a PSR requirement, which this package follows.

mikemand commented on config/config.php:
```diff
@@ -186,4 +188,24 @@
+            'class' => FileActivator::class,
+            'file' => storage_path('installed_modules'),
+            'cache-key' => 'activator.installed',
+            'cache-lifetime' => 604800
```
Needs trailing commas.

ryssbowh commented:
Yes, the cache lifetime for the file activator is in seconds.

Trailing commas... don't know when my brain will get used to that :)

nWidart commented on src/Activators/FileActivator.php:
```diff
@@ -0,0 +1,175 @@
+	}
+
+	/**
+	 * Get installed cache
```
This comment looks a bit unclear to me. From the comment alone, I expected this to return laravel's cache implementation.

nWidart commented on src/Activators/FileActivator.php:
```diff
@@ -0,0 +1,175 @@
+	 * 
+	 * @var array
+	 */
+	protected $installed;
```
Could this be renamed to `$installedModules`. I was hesitating if this was a boolean or not. 

Also what's an "installed module"? Is it an active module? If so, it might be more clear to use `activeModules` instead.

ryssbowh replied:
The way I see it, an installed module is a module that exists on the disk, it then has an activation status (1 or 0), both are stored by this activator.

nWidart replied:
Hm, thanks that explains the keyword a bit. So this contains an array of all modules? If so I'd still rename it `installedModules` to be explicit.

nWidart commented on src/Activators/FileActivator.php:
```diff
@@ -0,0 +1,175 @@
+
+class FileActivator implements ActivatorInterface
+{
+	/**
```
Why are fields/methods protected instead of private? Could you make all `protected` fields and methods private?

nWidart commented on src/Activators/FileActivator.php:
```diff
@@ -0,0 +1,175 @@
+	 * 
+	 * @return array
+	 */
+	protected function getCached()
```
Maybe the method could be named `getCachedModules`?

It seems like this doesn't always returns cached modules. If the cache is disabled, it won't from the looks of it, which doesn't seem to be the responsibility of this method, based on its name.

Actually, the only usage of this method is the constructor, so maybe this shouldn't mention cached at all and just be called `getInstalledModules` or `getActiveModules`.

nWidart commented on src/Activators/FileActivator.php:
```diff
@@ -0,0 +1,175 @@
+	}
+
+	/**
+	 * Forgets the installed cache
```
Comment seems weird to me, we don't delete cache, we delete the contents of the cache for a specific key in this implementation. 
I would call this `flushCache` or `clearCache`.

nWidart commented on src/Activators/FileActivator.php:
```diff
@@ -0,0 +1,175 @@
+		$this->cache = $app['cache'];
+		$this->files = $app['files'];
+		$this->config = $app['config'];
+		$this->fileInstalled = $this->config('file');
```
`fileInstalled` seems weird. Also the fact that there's a difference between the config key and the class field means there's something off with the naming. 
It should be the same in config key and class field. 

In order to chose a good name, What is this file? What does it contain? 
It looks like it contains a list of all active modules? 
If so we should use that keyword and avoid "install"

nWidart commented on src/Activators/FileActivator.php:
```diff
@@ -0,0 +1,175 @@
+    }
+
+	/**
+	 * Reads the installed json file
```
Can a json file be uninstalled? I think this read the json containing the active modules?

nWidart commented on src/Activators/FileActivator.php:
```diff
@@ -0,0 +1,175 @@
+	}
+
+	/**
+	 * Writes the installed json file
```
Same here, I wouldn't use "install"

nWidart commented on src/Activators/FileActivator.php:
```diff
@@ -0,0 +1,175 @@
+    public function isStatus(Module $module, $status)
+    {
+    	if(!isset($this->installed[$module->getName()])) return false;
+    	return $this->installed[$module->getName()] == $status;
```
Could you use strict comparison here, please? `===`

nWidart commented on src/Activators/FileActivator.php:
```diff
@@ -0,0 +1,175 @@
+    /**
+     * @inheritDoc
+     */
+    public function isStatus(Module $module, $status)
```
What's the type of the second argument, let's typehint this, I assume it's a boolean.

nWidart commented on src/Activators/FileActivator.php:
```diff
@@ -0,0 +1,175 @@
+    /**
+     * @inheritDoc
+     */
+    public function delete(Module $module)
```
This doesn't delete the module right? So let's call it `disableModule`?
Actually, I see there's also a disable method? So why would we ever want to delete it?

nWidart commented on src/Commands/ModuleMakeCommand.php:
```diff
@@ -35,9 +35,11 @@ public function handle()
                 ->setFilesystem($this->laravel['files'])
                 ->setModule($this->laravel['modules'])
                 ->setConfig($this->laravel['config'])
+                ->setActivator($this->laravel['modules.activator'])
```
Let's try to avoid the alias key and instead use the class reference. I'm trying to move away from the aliases.

nWidart commented on src/Contracts/ActivatorInterface.php:
```diff
@@ -0,0 +1,55 @@
+     * 
+     * @param  Module $module
+     */
+    public function delete(Module $module);
```
There's delete and disable? What's the difference?

ryssbowh replied:
This was only to make up for the delete method of the Module class. I assumed it's just a helper for developers or tests since I don't see any reference of it (?).
It basically deletes any activation status for that module (active or not)

nWidart replied:
I see ok, that makes sense. I believe the delete method on the module class is there by request from the community, there could a future delete command that uses that too actually.

nWidart commented on src/Generators/ModuleGenerator.php:
```diff
@@ -97,6 +114,20 @@ public function setPlain($plain)
+     *
+     * @return $this
+     */
+    public function setActive($active)
```
Let's typehint the argument to boolean?

ryssbowh replied:
I'd agree. But this argument comes from the Module class and it's not type hinted there, should I type hint it there too ?

nWidart replied:
Hm, this can still be type hinted without changing the Module class no? I think that would be a first step.
This package existed before scalar type hints in php so I'm only now moving towards them over time.

ryssbowh replied:
Ok, to be consistent with the Module class, I'll have to type hint them as integers even though they can only be 0 or 1. And they will have to be changed later, is that what you mean ?

nWidart replied:
Indeed, now that I think of it, maybe there's no need to have them as int, as there's no other status other than 0 or 1. Meaning the could be booleans.
If possible you can typehint on both sides, if not, don't worry I'll do that later on myself.

nWidart commented on src/Generators/ModuleGenerator.php:
```diff
@@ -97,6 +114,20 @@ public function setPlain($plain)
+     */
+    public function setActive($active)
+    {
+        $this->active = $active;
```
Since this is a boolean, please rename to `isActive`.

nWidart commented on src/Generators/ModuleGenerator.php:
```diff
@@ -131,6 +162,20 @@ public function setConfig($config)
+     *
+     * @return $this
+     */
+    public function setActivator($activator)
```
You can typehint the interface here too.

nWidart commented on src/LaravelModulesServiceProvider.php:
```diff
@@ -52,6 +52,13 @@ protected function registerServices()
+            return new $class($app);
+        });
         $this->app->alias(Contracts\RepositoryInterface::class, 'modules');
+        $this->app->alias(Contracts\ActivatorInterface::class, 'modules.activator');
```
I wouldn't create the alias here, I'm trying to move away from this practice.

nWidart commented:
This is a breaking change so it'll have to be for the next major release.

nWidart commented:
It looks like there's no tests at all for this new FileActivate class, this should be added.

ryssbowh commented:
Quite struggling with the tests as many tests (FileRepository, Module...) change module's activation statuses now, making the next tests assertions fail as the statuses are written in a separate file that is common to all tests, what would be the way to go you think :

- Each test that by its actions create a statuses file should delete it when it ends ?
- Or the following tests should take into account the previous tests activation statuses ?
- something else ?

nWidart commented:
Each test should be independent of one another so I think deleting the file after each test should be good.

ryssbowh commented:
A few things about the last 2 commits : 

- removed tests/stubs/valid/DisabledModule as status is not stored within the module folder anymore
- removed all statuses within tests/*/module.json

- I've not tested all the Lumen parts of these commits

- added a reset method to ActivatorInterface as I needed to clear data after the tests
- changed all statuses typehints to bools

nWidart commented:
Thank you for your patience.
Thank you for your contribution!

nWidart commented:
We should document this change in the docs, I'm confused now how to activate a module.

nWidart commented:
It looks like we need a command to migrate from the old way to handle active status, and write them in the activator file.

ryssbowh commented:
The way to activate a module hasn't changed, through the Module\enable() and \Module\disable().

Good point with the command, I'll write this asap.

nWidart commented:
I see I was wondering because it didn't detect old modules but that's indeed because they're not yet added to the new file.

I'm working on the migration command currently.

nWidart commented:
The current command can be viewed here: https://github.com/nWidart/laravel-modules/commit/a70af0dc04e26cc4f34bccdb08bb30591c9e155a

Let me know if I forgot something.

ryssbowh commented:
The only thing I can think of is that this command will fire 'enabling' and 'enable' events for each module, which you might not want to happen given that the modules are not technically enabled.

Maybe you can fetch the file used by the activator `$activator->getStatusesFilePath()` and write a statuses array 
```
[
  "Core" => true,
  "Media" => false
]
```
as json inside it.

And clear the activator cache afterwards. The key is `config('modules.activators.file.cache-key')`

ryssbowh commented:
Also,

It might be better to do this work within the activator class, in case someone writes a database handled activator that would need a different fix to this problem

nWidart commented:
Hm not sure if the events being fired is an actual issue. Since the events for enabled/disabled weren't really used before I doubt they do anything critical in user's codebases.

I'm not sure about doing this in the activator class, as it will have to know from which version to which version which probably isn't his responsibility. Maybe another class though.
